File: src/main/java/com/cxxcxx/zinecraft/core/dimension/StarGateTeleporter.java
package com.cxxcxx.zinecraft.core.dimension;

import com.cxxcxx.zinecraft.core.block.ModBlock;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction.Axis;
import net.minecraft.core.Vec3i;
import net.minecraft.resources.ResourceKey;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.LevelAccessor;
import net.minecraft.world.level.border.WorldBorder;
import net.minecraft.world.level.dimension.DimensionType;
import net.minecraft.world.level.levelgen.Heightmap.Types;
import net.minecraft.world.level.portal.DimensionTransition;
import net.minecraft.world.phys.Vec3;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public final class StarGateTeleporter {
  @NotNull
  public static final StarGateTeleporter INSTANCE = new StarGateTeleporter();

  private StarGateTeleporter() {
  }

  @Nullable
  public final DimensionTransition destination(@NotNull ServerLevel source, @NotNull Entity entity, @NotNull BlockPos portalPos) {
    ResourceKey target = source.dimension();
    ResourceKey resourceKey2;
    if (java.util.Objects.equals(target, Level.OVERWORLD)) {
      resourceKey2 = ModDimensions.TERRA.getLevelKey();
    } else {
      if (!java.util.Objects.equals(target, ModDimensions.TERRA.getLevelKey())) {
        return null;
      }

      resourceKey2 = Level.OVERWORLD;
    }

    ResourceKey resourceKey = resourceKey2;
    ServerLevel serverLevel1 = source.getServer().getLevel(resourceKey);
    if (serverLevel1 == null) {
      return null;
    }

    ServerLevel serverLevel = serverLevel1;
    double d = DimensionType.getTeleportationScale(source.dimensionType(), serverLevel.dimensionType());
    WorldBorder worldBorder1 = serverLevel.getWorldBorder();
    WorldBorder worldBorder = worldBorder1;
    BlockPos blockPos = worldBorder.clampToBounds(portalPos.getX() * d, portalPos.getY(), portalPos.getZ() * d);
    BlockPos blockPos1 = this.findGate(serverLevel, blockPos);
    BlockPos blockPos3 = blockPos1;
    if (blockPos1 == null) {
      blockPos3 = java.util.Objects.equals(serverLevel.dimension(), ModDimensions.TERRA.getLevelKey())
          ? this.createTerraGate(serverLevel, blockPos)
          : serverLevel.getSharedSpawnPos();
    }

    BlockPos blockPos2 = blockPos3;
    return new DimensionTransition(
        serverLevel,
        Vec3.atBottomCenterOf((Vec3i) blockPos2).add(0.0, 0.15, 0.0),
        Vec3.ZERO,
        entity.getYRot(),
        entity.getXRot(),
        DimensionTransition.PLAY_PORTAL_SOUND.then(DimensionTransition.PLACE_PORTAL_TICKET)
    );
  }

  private final BlockPos findGate(ServerLevel level, BlockPos around) {
    int i = around.getX() - 20;
    int j = around.getX() + 20;
    if (i <= j) {
      while (true) {
        int k = around.getZ() - 20;
        int l = around.getZ() + 20;
        if (k <= l) {
          while (true) {
            int m = level.getHeight(Types.MOTION_BLOCKING_NO_LEAVES, i, k);
            int n = Math.max(m - 10, level.getMinBuildHeight());
            int o = Math.min(m + 10, level.getMaxBuildHeight() - 1);
            if (n <= o) {
              while (true) {
                BlockPos blockPos = new BlockPos(i, n, k);
                if (level.getBlockState(blockPos).is(ModBlock.STARGATE_PORTAL.getBlock())) {
                  return blockPos;
                }

                if (n == o) {
                  break;
                }

                n++;
              }
            }

            if (k == l) {
              break;
            }

            k++;
          }
        }

        if (i == j) {
          break;
        }

        i++;
      }
    }

    return null;
  }

  private final BlockPos createTerraGate(ServerLevel level, BlockPos around) {
    BlockPos blockPos = level.getHeightmapPos(Types.MOTION_BLOCKING_NO_LEAVES, around);
    StarGateStructure starGateStructure = StarGateStructure.INSTANCE;
    LevelAccessor levelAccessor = (LevelAccessor) level;
    return starGateStructure.place(levelAccessor, blockPos, Axis.X, true);
  }
}

File: src/main/kotlin/com/cxxcxx/zinecraft/core/dimension/StarGateTeleporter.kt
package com.cxxcxx.zinecraft.core.dimension

import com.cxxcxx.zinecraft.core.block.ModBlock
import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.server.level.ServerLevel
import net.minecraft.world.entity.Entity
import net.minecraft.world.level.Level
import net.minecraft.world.level.border.WorldBorder
import net.minecraft.world.level.dimension.DimensionType
import net.minecraft.world.level.levelgen.Heightmap
import net.minecraft.world.level.portal.DimensionTransition
import net.minecraft.world.phys.Vec3

/** 星门的双向寻址、出口创建和安全落点辅助。 */
object StarGateTeleporter {
  fun destination(source: ServerLevel, entity: Entity, portalPos: BlockPos): DimensionTransition? {
    val targetKey = when (source.dimension()) {
      Level.OVERWORLD -> ModDimensions.TERRA.levelKey
      ModDimensions.TERRA.levelKey -> Level.OVERWORLD
      else -> return null
    }
    val target = source.server.getLevel(targetKey) ?: return null
    val scale = DimensionType.getTeleportationScale(source.dimensionType(), target.dimensionType())
    val border: WorldBorder = target.worldBorder
    val scaled = border.clampToBounds(portalPos.x * scale, portalPos.y.toDouble(), portalPos.z * scale)

    val existing = findGate(target, scaled)
    val arrival = when {
      existing != null -> existing
      target.dimension() == ModDimensions.TERRA.levelKey -> createTerraGate(target, scaled)
      else -> target.sharedSpawnPos
    }

    return DimensionTransition(
      target,
      Vec3.atBottomCenterOf(arrival).add(0.0, 0.15, 0.0),
      Vec3.ZERO,
      entity.yRot,
      entity.xRot,
      DimensionTransition.PLAY_PORTAL_SOUND.then(DimensionTransition.PLACE_PORTAL_TICKET)
    )
  }

  private fun findGate(level: ServerLevel, around: BlockPos): BlockPos? {
    for (x in around.x - 12..around.x + 12) {
      for (z in around.z - 12..around.z + 12) {
        val surface = level.getHeight(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, x, z)
        for (y in (surface - 10).coerceAtLeast(level.minBuildHeight)..(surface + 10).coerceAtMost(level.maxBuildHeight - 1)) {
          val pos = BlockPos(x, y, z)
          if (level.getBlockState(pos).`is`(ModBlock.STARGATE_PORTAL)) return pos
        }
      }
    }
    return null
  }

  private fun createTerraGate(level: ServerLevel, around: BlockPos): BlockPos {
    val base = level.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, around)
    return StarGateStructure.place(level, base, Direction.Axis.X)
  }
}
